python/map_closures/tools/visualizer.py

```diff
@@ -80,323 +109,482 @@ def update_closures(self, *kwargs):
+                "loop closures",
+                np.array(self._trajectory)[:, :3, -1],
+                np.array(self.loop_closures_data.closure_edges),
+                color=RED,
             )
-        else:
-            self.registration_vis.remove_geometry(
-                self.registration_data.source, reset_bounding_box=False
+            closure_lines.set_radius(0.1, relative=False)
+
+    def _unregister_trajectory(self):
+        self._ps.remove_point_cloud("trajectory")
+        if self.loop_closures_data.num_closures != 0:
+            self._ps.remove_curve_network("loop closures")
+
+    def _switch_view_callback(self):
+        if self.loop_closures_data.num_closures != 0:
+            self._gui.Separator()
+            BUTTON_NAME = (
+                REGISTRATION_VIEW_BUTTON if self._states.view_closures else CLOSURES_VIEW_BUTTON
             )
-        return False
-
-    def _toggle_keypoints(self, vis):
-        if self.render_source:
-            self.render_source = False
-            self.render_keypoints = True
-            self.registration_vis.remove_geometry(
-                self.registration_data.source, reset_bounding_box=False
+            if self._gui.Button(BUTTON_NAME) or self._gui.IsKeyPressed(self._gui.ImGuiKey_M):
+                self._states.view_closures = not self._states.view_closures
+                if self._states.view_closures:
+                    self._states.view_frame = False
+                    self._states.view_local_map = False
+                    self._states.view_closure_query = True
+                    self._states.view_closure_reference = True
+                    self._ps.get_point_cloud("current_frame").set_enabled(self._states.view_frame)
+                    self._ps.get_point_cloud("local_map").set_enabled(self._states.view_local_map)
+                    self._states.play_mode = False
+                    self._render_closure(self.loop_closures_data.current_closure_id)
+                    if not self._states.global_view:
+                        self._register_trajectory()
+                else:
+                    self._states.view_frame = True
+                    self._states.view_local_map = True
+                    self._states.view_closure_query = False
+                    self._states.view_closure_reference = False
+                    self._render_closure(self.loop_closures_data.current_closure_id)
+                    self._ps.get_point_cloud("current_frame").set_enabled(self._states.view_frame)
+                    self._ps.get_point_cloud("local_map").set_enabled(self._states.view_local_map)
+                    if not self._states.global_view:
+                        self._unregister_trajectory()
+                    self._ref_density_viewer = None
+                    self._query_density_viewer = None
+                    if self._states.view_density_map:
+                        self._states.view_density_map = False
+                        plt.close("all")
```

Comment:
I don't know why but this close is not working on my machine.

Reply:
This is a pain in the *** due to matplotlib and its blocking feature. This works with the latest matplotlib version only, which is mentioned in the optional deps in `pyproject.toml`.
We should replace matplotlib with some other Image visualizer, one option being opencv.